src/postprocess/refine.jl

```diff
@@ -0,0 +1,144 @@
+
+    # Virtual orbitals must be removed
+    ψ, occ = select_occupied_orbitals(basis, scfres.ψ, scfres.occupation; threshold=occ_threshold)
+    ψr = transfer_blochwave(ψ, basis, basis_ref)
+    ρr = transfer_density(scfres.ρ, basis, basis_ref)
+    _, hamr = energy_hamiltonian(basis_ref, ψr, occ; ρ=ρr)
+
+    # Compute the residual R(P)
+    # This recomputes ρr internally, but it's only a minor inefficiency.
+    res = compute_projected_gradient(basis_ref, ψr, occ)
+
+    # Compute M^{-1} R(P), with M^{-1} defined in [CDKL2021]
+    P = [PreconditionerTPA(basis_ref, kpt) for kpt in basis_ref.kpoints]
+    map(zip(P, ψr)) do (Pk, ψk)
+        precondprep!(Pk, ψk)
+    end
+
+    function apply_M(φk, Pk, δφnk, n)
+        proj_tangent_kpt!(δφnk, φk)
+        δφnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δφnk
+        proj_tangent_kpt!(δφnk, φk)
+        δφnk = sqrt.(Pk.mean_kin[n] .+ Pk.kin) .* δφnk
+        proj_tangent_kpt!(δφnk, φk)
+    end
+
+    function apply_inv_M(φk, Pk, δφnk, n)
+        proj_tangent_kpt!(δφnk, φk)
+        op(x) = apply_M(φk, Pk, x, n)
+        function f_ldiv!(x, y)
+            x .= proj_tangent_kpt(y, φk)
+            x ./= (Pk.mean_kin[n] .+ Pk.kin)
+            proj_tangent_kpt!(x, φk)
+        end
+        J = LinearMap{eltype(φk)}(op, size(δφnk, 1))
+        δφnk = IterativeSolvers.cg(J, δφnk, Pl=FunctionPreconditioner(f_ldiv!),
+                  verbose=false, reltol=0, abstol=1e-15)
+        proj_tangent_kpt!(δφnk, φk)
+    end
+
+    function apply_metric(φ, P, δφ, A::Function)
+        map(enumerate(δφ)) do (ik, δφk)
+            Aδφk = similar(δφk)
+            φk = φ[ik]
+            for n = 1:size(δφk,2)
+                Aδφk[:,n] = A(φk, P[ik], δφk[:,n], n)
+            end
+            Aδφk
+        end
+    end
+
+    # Compute the projection of the residual onto the high and low frequencies
+    resLF = transfer_blochwave(res, basis_ref, basis)
+    resHF = res - transfer_blochwave(resLF, basis, basis_ref)
+
+    # - Compute M^{-1}_22 R_2(P)
+    e2 = apply_metric(ψr, P, resHF, apply_inv_M)
+
+    # Apply Ω+K to M^{-1}_22 R_2(P)
+    Λ = map(enumerate(ψr)) do (ik, ψk)
+        Hk = hamr.blocks[ik]
+        Hψk = Hk * ψk
+        ψk'Hψk
+    end # Rayleigh coefficients
+    ΩpKe2 = apply_Ω(e2, ψr, hamr, Λ) .+ apply_K(basis_ref, e2, ψr, ρr, occ)
+    ΩpKe2 = transfer_blochwave(ΩpKe2, basis_ref, basis)
+
+    rhs = resLF - ΩpKe2
+
+    # Invert Ω+K on the small space
+    e1 = solve_ΩplusK(basis, ψ, rhs, occ; tol=ΩpK_tol).δψ
+
+    e1 = transfer_blochwave(e1, basis, basis_ref)
+    schur_residual = e1 + e2
+
+    # Use the Schur residual to compute (minus) the first-order correction to
+    # the density.
+    δρ = compute_δρ(basis_ref, ψr, schur_residual, occ)
+
+    RefinementResult(basis_ref, ψr, ρr, occ, schur_residual, δρ)
+end
+
+"""
+Retrieve the refined density from a [`RefinementResult`](@ref).
+"""
+function refine_density(refinement::RefinementResult)
```

Comment:
I'm wondering if it may be better to target `-δρ` (i.e. the correction) instead of `ρ - δρ` as the quantity to return. Pros is that you will need `δρ` as an error estimate and that doing the final manipulation of adding the correction and `ρ` is trivial and obvious.

Reply:
But forcing to compute a quantity and its correction at the same time has advantages too as one can use tricks such as `DiffResults` to avoid duplicate computations, see below.

Reply:
I changed it to return both `F` and `\deltaF` for forces. (And `refine_density` is gone since `\rho` and `\delta\rho` are both in the refinement info already).

DiffResults support will be for another time.